avocado/utils/vmimage.py

```diff
@@ -0,0 +1,375 @@
+        # corresponding format and we will get the highest
+        # occurrence of that expression.
+        self.default_version = '[0-9]+.[0-9]+'  # 16.10/17.04/...
+        self.default_build = None  # Ubuntu doesn't seem to mark builds
+        self.default_arch = 'amd64'
+
+        # Dictionary of expected arguments and their values
+        self.args = {'version': kwargs.get('version', self.default_version),
+                     'build': kwargs.get('build', self.default_build),
+                     'arch': kwargs.get('arch', self.default_arch),
+                     'checksum': kwargs.get('checksum', None),
+                     'algorithm': kwargs.get('algorithm', None),
+                     'cache_dir': kwargs.get('cache_dir',
+                                             tempfile.gettempdir())}
+
+
+class DebianImageProvider(ImageBase):
+
+    name = 'Debian'
+
+    def __init__(self, **kwargs):
+        super(DebianImageProvider, self).__init__(**kwargs)
+        # Here you can use string formatting tags corresponding to the
+        # arguments accepted (version, build and arch).
+        self.url_versions = 'https://cdimage.debian.org/cdimage/openstack/'
+        self.url_images = self.url_versions + '/{version}/'
+        self.image = 'debian-{version}-openstack-{arch}.qcow2$'
+
+        # Set the defaults to be used when the arguments are
+        # not provided. If you don't want to force a version,
+        # you can provide a regular expression that matches the
+        # corresponding format and we will get the highest
+        # occurrence of that expression.
+        self.default_version = '[0-9]+.[0-9]+.[0-9]+-.*'  # 9.1.4-20170830
+        self.default_build = None  # Build is part of the image version
+        self.default_arch = 'amd64'
+
+        # Dictionary of expected arguments and their values
+        self.args = {'version': kwargs.get('version', self.default_version),
+                     'build': kwargs.get('build', self.default_build),
+                     'arch': kwargs.get('arch', self.default_arch),
+                     'checksum': kwargs.get('checksum', None),
+                     'algorithm': kwargs.get('algorithm', None),
+                     'cache_dir': kwargs.get('cache_dir',
+                                             tempfile.gettempdir())}
+
+
+class ImageProviderProxy(object):
+    def __init__(self):
+        self.providers_list = []
+        self.providers_list.append(FedoraImageProvider)
+        self.providers_list.append(FedoraSecondaryImageProvider)
+        self.providers_list.append(CentOSImageProvider)
+        self.providers_list.append(UbuntuImageProvider)
+        self.providers_list.append(DebianImageProvider)
+
+    def get(self, name=None, **kwargs):
+        # No name, return the first registered provider
+        if name is None and self.providers_list:
+            return self.providers_list[0](**kwargs).get()
```

Comment:
Well in case the name is `None` I'd expect it to ask all the providers whether they already have such image and only when no image available I'd return the first one.

Basically

```python
get("Ubuntu")
get()  => "Ubuntu"
get("CentOs", version=5)
get() => "CentOs"
get("Fedora")
get() => "Fedora"
get(None, version=5) => CentOS
```

Reply:
Anyway that is just RFE, not necessarily now... (only the API should allow such expansion)

Reply:
Yes, good point. This is easy implementation. Including this in the next version.